### PR TITLE
[Numpy] Numpy vstack

### PR DESCRIPTION
**Pull request by hzfan**
## Description ##
Numpy compatible vstack with existing kernel from concat
[vstack](https://docs.scipy.org/doc/numpy/reference/generated/numpy.vstack.html)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- add stack internal interface
- add python interface
- add unit tests

## Comments ##
Thank @reminisce @haojin2 for guidance and review.

### CODE REVIEW

File: src/operator/numpy/np_matrix_op.cc [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  }
+  for (int i = 0; i < param.num_args; i++) {
+    TYPE_ASSIGN_CHECK(*in_type, i, dtype);
+  }
+  TYPE_ASSIGN_CHECK(*out_type, 0, dtype);
+  return dtype != -1;
+}
+
+bool NumpyVstackShape(const nnvm::NodeAttrs& attrs,
+                            mxnet::ShapeVector* in_attrs,
```
Comment (haojin2) [inline src/operator/numpy/np_matrix_op.cc]: Alignment of arguments

File: python/mxnet/numpy/multiarray.py [hunk 2]
```diff
@@ -44,7 +44,7 @@
 from ..ndarray.numpy import _internal as _npi
 
 __all__ = ['ndarray', 'empty', 'array', 'zeros', 'ones', 'add', 'subtract', 'multiply', 'divide',
-           'mod', 'power']
+           'mod', 'power', 'vstack']
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py]: after rebase, put `vstack` to after `stack` in this list and within the file, same rule applies to the other 2 files too.

**Comment by szha** [comment]
@hzfan could you resolve conflict?

**Comment by hzfan** [comment]
@szha conflicts resolved.

File: src/operator/numpy/np_matrix_op.cc (hunk 1 above)
Comment (hzfan) [reply src/operator/numpy/np_matrix_op.cc]: Aligned.

File: python/mxnet/numpy/multiarray.py (hunk 2 above)
Comment (hzfan) [reply python/mxnet/numpy/multiarray.py]: Done.

**Pull request merged**